File: Python/Planners/Graphbased/ImprovedStateLattice/state_lattice_planner.py
#! ./venv/bin/python
from functools import partial
import heapq

import numpy as np
import globalvar
from state_lattice import StateLatticeGraph, ActionSet
from main_unstructure import GenerateStaticObstacles_unstructured, VisualizeDynamicResults, \
    checkObj_linev, checkObj_point
from TransformPathToTrajectory import TransformPathToTrajectory
import json

DEBUG = False

class LUT:
    """
        look up table 
        use load_LUT(...) to load file from disk.
        use look_up(...) to get the value
    """
    def __init__(self) -> None:
        self.lut = None
        self.edge_length = None
        self.direction_num = 8

    def load_LUT(self, filename: str):
        """
            load lookup table file from disk
        Args:
            filename: str, filename of the lut file

        """
        f = open(filename, 'r', encoding='utf8')
        json_str = f.read()
        record = json.loads(json_str)
        edge_length = record['edge_length']

        look_up_table = []
        for direction in range(8):
            look_up_table.append(record[f"{direction}"])
        
        self.lut, self.edge_length = np.array(look_up_table), edge_length

    def look_up(self, start_direction, to_pos, to_direction):
        """
            get the value of specific pos and direction.
        Args:
            start_direction: integer in 0~7 which refers to theta of direction * pi/4,
                It's the initial direction of the vehicle.
            to_pos: tuple of (int, int). position of the destination. 
                x, y must in [-1, 0, 1].
            to_direction: integer in 0~7, the expect direction(or theta) when car reach destination.
        
        Returns:
            the expected cost of the given action.
        
        All these params control the moving method of the vehicle.

        """
        # default start at pos (0, 0)
        x, y = to_pos
        assert x < self.edge_length and y < self.edge_length, f"x: {x}, y: {y}"
        index = x * self.edge_length * self.direction_num + y * self.direction_num + to_direction

        return self.lut[start_direction, index]


def theta2direction(theta):
    """
        convert theta to direction:
    Args:
        theta: float, the degree of direction in radian.

    Returns:
        the direction of 0~7
    """
    return np.round(theta / (np.pi / 4)).astype(int) % 8


def direction2theta(direction):
    """
        convert direction to theta:
    Args:
        theta: float, the degree of direction in radian.

    Returns:
        theta in radian.
    """
    return direction * np.pi / 4


def loc2pos(loc, scale):
    """
        Convert location to position. In this context, loc means absolute position 
        in environment, pos means the relative position in graph.
    Args:
        loc: (float, float). The absolute position in planning_scale_
        scale: float. the scaling factor of graph and planning_scale_

    Returns:
        the pos if type of (int , int)
    """
    planning_scale_ = globalvar.planning_scale_
    x_min = planning_scale_.xmin
    y_min = planning_scale_.ymin
    return int((loc[0] - x_min) / scale), int((loc[1] - y_min)/scale)


def pos2loc(pos, scale):
    """
        Convert position to location. In this context, loc means absolute position 
        in environment, pos means the relative position in graph.
    Args:
        pos: (int, int). The absolute position in planning_scale_
        scale: float. the scaling factor of graph and planning_scale_

    Returns:
        the loc if type of (float , float)
    """
    pos = np.array(pos)
    planning_scale_ = globalvar.planning_scale_
    x_min = planning_scale_.xmin
    y_min = planning_scale_.ymin
    if len(pos.shape) == 1:
        return pos[0]* scale + x_min, scale * pos[1] + y_min
    else:
        pos[:, 0] = pos[:, 0] * scale + x_min
        pos[:, 1] = pos[:, 1] * scale + y_min
        return pos


def Astar(graph: StateLatticeGraph, start_pos, start_direction, end_pos, end_direction, heuristic_func, pos_validate):
    """
        A* search method. This function will give a possible way from start to end. The problem is defined 
        in an searh problem in StateLatticeGraph.
    Args:
        graph: instance of StateLatticeGraph. Graph contains necessary information for calculation.
        start_pos: (int, int). The start position of the vehicle. Must inside the graph.
        start_direction: integer of 0~7. The start direction of the vehicle.
        end_pos: (int, int). The destination.
        end_direction: integer of 0~7. The direction when finish moving.
        heuristic_func: python function. This function will be called like 
            heuristic_func(graph, adj_pos, adj_direction, end_pos, end_direction)
        pos_validate: python function. This function will be called like
            pos_validate(graph, current_pos, current_direction, other_pos, other_direction):
        
    Returns:
        (g_score, path)
        g_score: list[int]. The actual cost of the point. Each index of g_score refers to a specific (pos, direction).
            The pos ans direction can be extract by calling graph.index2position(index) 
        path: dict. This dict contains the node index and its predecessor in the best way.

    """
    nodes = graph.nodes
    edges = graph.edges

    start_index = graph.position2index(start_pos, start_direction)
    end_index = graph.position2index(end_pos, end_direction)

    # start A*
    close_index = []
    open_list = []
    path = {}

    # initialize the start node, calculate f = g + h
    g_score = np.full((len(nodes), ), float('inf'))
    g_score[start_index] = 0

    h_start = heuristic_func(graph, start_pos, start_direction, end_pos, end_direction)
    f_start = g_score[start_index] + h_start
    heapq.heappush(open_list, (f_start, start_index))

    while len(open_list) != 0:
        # get the node with the minimum f, move it to close_index
        current_f, current_index = heapq.heappop(open_list)
        current_pos, current_direction = graph.index2position(current_index)
        close_index.append(current_index)

        # if end_index in open_list and it has the best f score,
        # we can end the searching procedure
        if current_index == end_index:
            break

        for adj_index, edge_cost in edges[current_index]:
            if adj_index in close_index:
                continue

            # if there is obstacle in edge, it means the adj_index is actually
            # not a adjacent node of current node. So we continue and ignore it.
            adj_pos, adj_direction = graph.index2position(adj_index)
            if not pos_validate(graph, current_pos, current_direction, adj_pos, adj_direction):
                continue
            temp_g_score = g_score[current_index] + edge_cost
            if temp_g_score < g_score[adj_index]:
                path[adj_index] = current_index
                g_score[adj_index] = temp_g_score

                f_score = g_score[adj_index] + heuristic_func(graph, adj_pos, adj_direction, end_pos, end_direction)
                heapq.heappush(open_list, (f_score, adj_index))
        
    return g_score, path


def check_move(graph, from_pos, from_direction, to_pos, to_direction, scale=1):
    """
        A simple function to judge whether there are obstacles between from_pos 
        and to_pos.
    Args:
        graph: instance of StateLatticeGraph.
        from_pos: (int ,int). The start position.
        from_direction: integer of 0~7. the start direction.
        to_pos: (int, int). The ending position.
        to_direction: integer of 0~7. The ending direction.
        scale: float. The scaling factor of planning_scale_ and position inside the graph.
    
    Returns:
        If move is valid, return True else False.
    """
    obstacles = globalvar.obstacles_

    from_loc = pos2loc(from_pos, scale)
    to_loc = pos2loc(to_pos, scale)

    for obs in obstacles[0]:
        if checkObj_point(np.array(to_loc), np.vstack([obs.x, obs.y])):
            return False
    
        if checkObj_linev(np.array(from_loc), np.array(to_loc), np.vstack([obs.x, obs.y])):
            return False
    
    return True


def hlut(lut, graph, from_pos, from_direction, to_pos, to_direction):
    """
        The Heuristic LookUp Table. An heuristic fuction implemented with lookup table.
    Args:
        lut: instance of LUT. The lookup table with params loaded.
        graph: instance of StateLatticeGraph.
        from_pos: (int ,int). The start position.
        from_direction: integer of 0~7. the start direction.
        to_pos: (int, int). The ending position.
        to_direction: integer of 0~7. The ending direction.
    
    Returns:
        The value of the guessing cost from (from_pos, from_direction) to (to_pos, to_direction)
    """
    edge_length = lut.edge_length
    delta_x, delta_y = np.array(to_pos) - np.array(from_pos)
    
    # if actual graph is bigger than lut,
    # we use an overflow factor to describe how bigger it is,
    # the result will be multiplied by this factor

    # because lut can not process negitive position, we 
    # need to reconstuct an euqal move which has positive position
    if delta_x < 0:
        delta_x = - delta_x
        # reflect with y-axis
        from_direction = ((4 - from_direction) + 8) % 8
        to_direction = ((4 - to_direction) + 8) % 8
    if delta_y < 0:
        delta_y = - delta_y
        # reflect with x-axis
        from_direction = ((- from_direction) + 8) % 8
        to_direction = ((- to_direction) + 8) % 8

    # if move is smaller than lut, then directly return the result
    if delta_x < edge_length and delta_y < edge_length:
        return lut.look_up(from_direction, (delta_x, delta_y), to_direction)
    
    # else we calculate the overload_factor, delta_x and delta_y
    if delta_x < edge_length and delta_y >= edge_length:
        overflow_factor = (delta_y + 1) / edge_length
        delta_x = np.round(delta_x * edge_length / (delta_y + 1)).astype(int)
        delta_y = edge_length - 1

    elif delta_x >= edge_length and delta_y < edge_length:
        overflow_factor = (delta_x + 1) / edge_length
        delta_y = np.round(delta_y * edge_length / (delta_x + 1)).astype(int)
        delta_x = edge_length - 1
    else:
        overflow_factor = (delta_x + 1) / (delta_y + 1) if delta_x > delta_y else (delta_y + 1) / (delta_x + 1)
        delta_x = min(np.round(edge_length * (delta_x + 1) / (delta_y + 1)).astype(int), edge_length - 1)
        delta_y = min(np.round(edge_length * (delta_y + 1) / (delta_x + 1)).astype(int), edge_length - 1)
    
    # return the estimation result.
    return overflow_factor * lut.look_up(from_direction, (delta_x, delta_y), to_direction)


def ImprovedStateLatticePlan(loc2pos_scale=1, lutfile="./look_up_table_121628.txt"):
    """
        Improved State Lattice Planner. This function will give
        a possible trajectory from start to end.
    Args:
        loc2pos_scale: float. Refers to edge scale between graph and actual map.

    Returns:
        [x, y, theta, path_length, is_complete] as defined above.
    """

    # initalize global variables
    obstacles = globalvar.obstacles_
    vehicle_TPBV_ = globalvar.vehicle_TPBV_
    planning_scale_ = globalvar.planning_scale_
    vehicle_geometrics_ = globalvar.vehicle_geometrics_
    vehicle_kinematics_ = globalvar.vehicle_kinematics_

    # process scaling. this will define how big is the graph.
    graph_width = planning_scale_.x_scale
    graph_height = planning_scale_.y_scale
    graph = StateLatticeGraph(int(graph_width / loc2pos_scale), int(graph_height / loc2pos_scale))

    # get the start location and the end location.
    start_loc, start_direction = (vehicle_TPBV_.x0, vehicle_TPBV_.y0), theta2direction(vehicle_TPBV_.theta0)
    end_loc, end_direction = (vehicle_TPBV_.xtf, vehicle_TPBV_.ytf), theta2direction(vehicle_TPBV_.thetatf)

    # convert location to position in graph.
    start_pos = loc2pos(start_loc, loc2pos_scale)
    end_pos = loc2pos(end_loc, loc2pos_scale)

    end_index = graph.position2index(end_pos, end_direction)
    start_index = graph.position2index(start_pos, start_direction)

    # initial lookup table
    lut = LUT()
    lut.load_LUT(lutfile)
 
    # start A* fucntion 
    start_time = time.time()
    g_score, path = Astar(graph, start_pos, start_direction, end_pos, end_direction, partial(hlut, lut), partial(check_move, scale=loc2pos_scale))
    time_consumption = time.time() - start_time
    print(f"Astar time comsumption: {time_consumption}")

    # reconstruct the minimum cost path.
    result = []
    i = end_index
    if DEBUG:
        print(f"{graph.nodes[i]}-{g_score[i]}", end='')
    result.append((graph.nodes[i].position, graph.nodes[i].direction))
    if i in path:
        current = path[i]
        while current in path:
            n = graph.nodes[current]
            if DEBUG:
                print(f' <- {n}-{g_score[current]}', end='')
            result.append((n.position, n.direction))
            current = path[current]
        if DEBUG:
            print(f' <- {graph.nodes[start_index]}-{g_score[start_index]}', end='')
        result.append((start_pos, start_direction))
    else:
        # if we can not go to the destination, just return 0 of is_complete.
        return [[0], [0], [0], 0, 0]
    print('')

    result.reverse()
    xy_list = np.array([r[0] for r in result])
    direction_list = np.array([r[1] for r in result])
    
    # The result is attained but it is represented only by the lattice vertex
    # in the graph. It is discrete and unfriendly to human.
    # We need to add more points in the path to make it smoother.
    xy_full = []
    theta_full = []
    for i in range(len(xy_list)-1):
        # ActionSet.sample_points can give more points base on the action
        # it will automatically extract action from pos.
        xy_points, theta_points = ActionSet.sample_points(xy_list[i], direction_list[i], xy_list[i+1], direction_list[i+1])
        xy_full.extend(xy_points)
        theta_full.extend(theta_points)
    xy_full_loc = pos2loc(xy_full, loc2pos_scale)
    
    return [xy_full_loc[:, 0], xy_full_loc[:, 1], theta_full, 
        g_score[end_index] * loc2pos_scale, 1] 


if __name__ == "__main__":
    import time

    globalvar.Nobs = 0

    globalvar.obstacles_ = GenerateStaticObstacles_unstructured()
    # print(globalvar.vehicle_kinematics_.min_turning_radius)
    # print(globalvar.vehicle_geometrics_.r2x)
    
    start_time = time.time()
    x, y, theta, path_length, is_complete = ImprovedStateLatticePlan(1, "./look_up_table_121628.txt")
    # print(theta)
    # print(len(x), len(y), len(theta), path_length)
    time_consumption = time.time() - start_time
    print(f"ImprovedStateLatticePlan time comsumption: {time_consumption}")
    transmethod_flag=1 # choose different path to trajectory method
    trajectory = TransformPathToTrajectory(x, y, theta, path_length, transmethod_flag)

    VisualizeDynamicResults(trajectory)

    global obs

File: Cpp/Planners/Graphbased/ImprovedStateLattice/state_lattice_planner.cpp
#include "state_lattice_planner.h"
#include <fstream>
#include <iostream>
#include "mathstruct.h"
#include <queue>
#include <algorithm>
#include <chrono>

using namespace std;

// global variables
extern Vehicle_geometrics_ vehicle_geometrics_;
extern Vehicle_kinematics_ vehicle_kinematics_;
extern Hybrid_astar_ hybrid_astar_;
extern Vehicle_TPBV_ vehicle_TPBV_;
extern Planning_scale_ planning_scale_;
extern vector<vector<math::Vec2d> > obstacles_;


void LUT::load_LUT(string filename)
{
    ifstream f;
    stringstream buffer;

    f.open(filename);
    assert(f.is_open());
    buffer << f.rdbuf();
    string s = buffer.str();

    json result = json::parse(s.data());
    // load the edge_length
    this->edge_length = result["edge_length"].get<int>();
    // load the lut data.
    this->lut = vector<vector<double>>(8);
    for(int i = 0; i < this->direction_num; i++)
    {
        this->lut[i] = result[to_string(i)].get<vector<double>>();
    }
}

double LUT::look_up(int start_direction, Position to_pos, int to_direction) const 
{
    int index = to_pos.x() * this->edge_length * this->direction_num + to_pos.y() * this->direction_num
        + to_direction;
    return this->lut[start_direction][index];
}

int theta2direction(double theta)
{
    return (int)round(theta / M_PI_4) % 8;
}

double direction2theta(int direction)
{
    return direction * M_PI_4;
}

Position loc2pos(Vec2d loc, double scale)
{
    double xmin = planning_scale_.xmin;
    double ymin = planning_scale_.ymin;

    return Position((int)floor((loc.x() - xmin) / scale), (int)floor((loc.y() - ymin) / scale) );
}

Vec2d pos2loc(Position pos, double scale)
{
    double xmin = planning_scale_.xmin;
    double ymin = planning_scale_.ymin;
    return Vec2d(pos.x() * scale + xmin, pos.y() *scale + ymin);
}

Vec2d pos2loc(Vec2d pos, double scale)
{
    double xmin = planning_scale_.xmin;
    double ymin = planning_scale_.ymin;
    return Vec2d(pos.x() * scale + xmin, pos.y() *scale + ymin);
}

pair<vector<double>, vector<NodeIndex>> Astar(const StateLatticeGraph & graph, 
    Position start_pos, int start_direction, Position end_pos, int end_direction,
    std::function<double(const StateLatticeGraph &, Position, int, Position, int)> heuristic_func,
    std::function<bool(const StateLatticeGraph &, Position, int, Position, int)> pos_validate
)
{
    // define the path. Each element is the previous node index of current node index
    auto path = vector<NodeIndex>(graph.node_num, -1);
    const auto & nodes = graph.nodes;
    const auto & edges = graph.edges;

    NodeIndex start_index = graph.position2index(start_pos, start_direction);
    NodeIndex end_index = graph.position2index(end_pos, end_direction);

    auto close_index = vector<NodeIndex>();
    // open_list is optimized by priority_queue
    priority_queue<pair<double, NodeIndex>> open_list;

    // calculate g_score, h_score, f_score = g_score + h_score.
    auto g_score = vector<double>(graph.node_num, numeric_limits<double>::infinity());
    g_score[start_index] = 0.0;
    auto h_score = heuristic_func(graph, start_pos, start_direction, end_pos, end_direction);
    auto f_start = h_score + g_score[start_index];

    // warning, remember to use -fscore in priority_queue
    // beacuse the priority_queue always pop the max
    open_list.push(make_pair(-f_start, start_index));

    while (!open_list.empty())
    {
        // pop the node with the min f_score.
        auto temp_record = open_list.top();
        open_list.pop();
        NodeIndex current_index = temp_record.second;
        auto temp_result = graph.index2position(current_index);
        Position current_pos = temp_result.first;
        int current_direction = temp_result.second;
        // put it to close list
        close_index.push_back(current_index);

        if (current_index == end_index)
        {
            break;
        }

        const auto & current_edges = edges[current_index];
        for (auto iter = current_edges.begin(); iter < current_edges.end(); iter++)
        {
            NodeIndex adj_index = iter->index;
            double edge_cost = iter->cost;
            
            // if adj_index in close_index then continue
            if (find(close_index.begin(), close_index.end(), adj_index) != close_index.end())
            {
                continue;
            }
            auto temp_result = graph.index2position(adj_index);
            auto adj_pos = temp_result.first;
            auto adj_direction = temp_result.second;

            // check if the move from current to adj is valid.
            if (!pos_validate(graph, current_pos, current_direction, adj_pos, adj_direction))
            {
                // if not valid. Then the adj is not connect to the current.
                // continue
                continue;
            }
            
            double temp_g_score = g_score[current_index] + edge_cost;
            if (temp_g_score < g_score[adj_index])
            {
                path[adj_index] = current_index;
                g_score[adj_index] = temp_g_score;

                auto f_score = g_score[adj_index] + 
                    heuristic_func(graph, adj_pos, adj_direction, end_pos, end_direction);
                // warning, remember to use -fscore in priority_queue
                open_list.push(make_pair(-f_score, adj_index));
            }

        }
        
    }
    return make_pair(g_score, path);

}

bool check_move(const StateLatticeGraph & graph, Position from_pos, int from_direction, 
    Position to_pos, int to_direction, double loc2pos_scale)
{
    auto from_loc = pos2loc(from_pos, loc2pos_scale);
    auto to_loc = pos2loc(to_pos, loc2pos_scale);
    vector<math::Vec2d> V = CreateVehiclePolygon(to_loc.x(), to_loc.y(), direction2theta(to_direction));
    cout << V.size() << endl;

    auto iter = obstacles_.begin();
    for(; iter < obstacles_.end(); iter++)
    {   
        // we assume that the from_pos is valid.
        if (checkObj_point(to_loc, *iter))
        {
            return false;
        }
        if (checkObj_linev(V[0], V[1], *iter)) {
            return false;
        }
        if (checkObj_linev(V[1], V[2], *iter)) {
            return false;
        }
        if (checkObj_linev(V[2], V[3], *iter)) {
            return false;
        }
        if (checkObj_linev(V[3], V[0], *iter)) {
            return false;
        }
        // TODO: it seems I can check all obstacles in one time.
        if (checkObj_linev(from_loc, to_loc, *iter))
        {
            return false;
        }
    }
    return true;
}

double hlut(const LUT & lut, const StateLatticeGraph & graph, Position start_pos, int start_direction, Position end_pos, int end_direction)
{
    auto edge_length = lut.edge_length;
    auto delta_pos = end_pos - start_pos;
    
    // if the delta_position is bigger then the lut,
    // the lut can not give an precise result.
    // we use overflow_facotr to estimate the result.
    double overflow_factor = 1;

    int delta_x = delta_pos.x(); 
    int delta_y = delta_pos.y();

    if (delta_x < 0)
    {
        // reflect with t-axis
        delta_x = - delta_x;
        start_direction = ((4 - start_direction) + 8) % 8;
        end_direction = ((4 - end_direction) + 8) % 8;
    } 

    if (delta_y < 0)
    {
        // reflect with x-axis
        delta_y = - delta_y;
        start_direction = ((- start_direction) + 8) % 8;
        end_direction = ((- end_direction) + 8) % 8;
    }
    
    // if delta_x < edge_length && delta_y < endge_length 
    // then directly return
    if (delta_x < edge_length && delta_y < edge_length)
    {
        return lut.look_up(start_direction, Position(delta_x, delta_y), end_direction);
    }

    // use Euclidean distance to esitimate
    return std::hypot(delta_x, delta_y);
}

void path_interpolation(PlanningResult & result, const vector<pair<Position, int>>& path_result_list,
    const ActionSet & action_set, double loc2pos_scale)
{
    // add more points to make trajectory smooth
    auto iter = path_result_list.begin();
    vector<vector<double>> theta_full;
    for(; iter < path_result_list.end() - 1; iter++)
    {   
        auto start_pos = iter->first;
        auto start_direction = iter->second;
        auto end_pos = (iter+1)->first;
        auto end_direction = (iter+1)->second;
        auto points_result = action_set.sample_points(start_pos, start_direction, end_pos, end_direction);
        
        // convert position to absolute location
        auto temp_x_result = points_result.first.first;
        auto temp_y_result = points_result.first.second;
        auto x_len = temp_x_result.size();
        for (int i = 0; i < x_len; i++)
        {
            Vec2d temp_pos = Vec2d(temp_x_result[i], temp_y_result[i]);
            Vec2d loc_res = pos2loc(temp_pos, loc2pos_scale);
            result.x.push_back(loc_res.x());
            result.y.push_back(loc_res.y());
        }
        auto temp_theta_result = points_result.second;
        result.theta.insert(result.theta.end(), temp_theta_result.begin(), temp_theta_result.end());
    }
}


PlanningResult ImprovedStateLatticePlan(double loc2pos_scale, string lut_file, string insert_points_file)
{
    // process scaling. this will define how big is the graph.
    double graph_width = planning_scale_.x_scale();
    double graph_height = planning_scale_.y_scale();
    auto action_set = ActionSet();
    action_set.load_data(insert_points_file);
    auto graph = StateLatticeGraph((int)(graph_width / loc2pos_scale), (int)(graph_height / loc2pos_scale), action_set);

    Vec2d start_loc = Vec2d(vehicle_TPBV_.x0, vehicle_TPBV_.y0);
    int start_direction = theta2direction(vehicle_TPBV_.theta0);
    Vec2d end_loc = Vec2d(vehicle_TPBV_.xtf, vehicle_TPBV_.ytf);
    int end_direction = theta2direction(vehicle_TPBV_.thetatf);

    // convert loc to pos
    Position start_pos = loc2pos(start_loc, loc2pos_scale);
    Position end_pos = loc2pos(end_loc, loc2pos_scale);

    NodeIndex start_index = graph.position2index(start_pos, start_direction);
    NodeIndex end_index = graph.position2index(end_pos, end_direction);

    LUT lut;
    lut.load_LUT(lut_file);
    // cout << "load successfully" << endl;

    // In C++, lambda function with variables captured can not be 
    // cast to a function pointer.
    // I have to use std::function to pass the closures.
    std::function<double(const StateLatticeGraph &, Position, int, Position, int)> heuristic_func = 
    [&](const StateLatticeGraph & g, Position s_p, int s_d, Position e_p, int e_d) 
    {
        // static int call_times = 0;
        // call_times++;
        // cout << "call hlut: " << call_times << endl;
        return hlut(lut, g, s_p, s_d, e_p, e_d);
    };
    std::function<bool(const StateLatticeGraph &, Position, int, Position, int)> position_check = 
    [&](const StateLatticeGraph & g, Position s_p, int s_d, Position e_p, int e_d)
    {
        return check_move(g, s_p, s_d, e_p, e_d, loc2pos_scale);
    };

    auto start_time = chrono::high_resolution_clock::now();
    auto astar_result = Astar(graph, start_pos, start_direction, end_pos, end_direction, heuristic_func, position_check);
    auto end_time = chrono::high_resolution_clock::now();
    std::cout << "Astar time consumed: " << 
        chrono::duration_cast<chrono::microseconds>(end_time - start_time).count() <<
        " us"
        << std::endl;
    auto g_score = astar_result.first;
    auto path = astar_result.second;

    vector<pair<Position, int>> temp_result_list;

    // construct the path from start to end.
    auto current_index = end_index;
    if (path[current_index] == -1)
    {
        return PlanningResult();
    }
    while(path[current_index] != -1)
    {
        temp_result_list.push_back(graph.index2position(current_index));
        current_index = path[current_index];
    }
    temp_result_list.push_back(graph.index2position(start_index));
    // reverse the path 
    std::reverse(temp_result_list.begin(), temp_result_list.end());

    PlanningResult result;

    path_interpolation(result, temp_result_list, action_set, loc2pos_scale);

    
    result.path_length = g_score[end_index] * loc2pos_scale;
    result.is_complete = 1;
    auto end_time2 = chrono::high_resolution_clock::now();
        std::cout << "Astar+interpolation time consumed: " << 
        chrono::duration_cast<chrono::microseconds>(end_time2 - start_time).count() <<
        " us"
        << std::endl;
    return result;
}
